velox/exec/tests/RowContainerTest.cpp

```diff
@@ -304,23 +346,147 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+                       l.second,
+                       rowContainer->columnAt(0),
+                       decoded,
+                       r.first,
+                       {nullsFirst, ascending}) < 0;
+          });
+      auto result =
+          createResultVectorForSimpleComparison<T>(indexedRows, decoded);
+      assertEqualVectors(expected, result);
+    }
+
+    // Verify compareRows method with row as input.
+    {
+      std::sort(
+          indexedRows.begin(),
+          indexedRows.end(),
+          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
+            return rowContainer->compareRows(
+                       l.second, r.second, {{nullsFirst, ascending}}) < 0;
+          });
+      auto result =
+          createResultVectorForSimpleComparison<T>(indexedRows, decoded);
+      assertEqualVectors(expected, result);
+    }
+
+    // Verify compareRows method with default compare flags.
+    if (ascending && nullsFirst) {
+      std::sort(
+          indexedRows.begin(),
+          indexedRows.end(),
+          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
+            return rowContainer->compareRows(l.second, r.second) < 0;
+          });
+      auto result =
+          createResultVectorForSimpleComparison<T>(indexedRows, decoded);
+      assertEqualVectors(expected, result);
+    }
+  }
+
+  template <typename T>
+  VectorPtr createResultVectorForArrayComparison(
+      const std::vector<std::pair<vector_size_t, char*>>& indexedRows,
+      const DecodedVector& decoded) {
+    std::vector<std::optional<std::vector<std::optional<T>>>> sorted;
+    for (const auto& irow : indexedRows) {
+      if (decoded.isNullAt(irow.first)) {
+        sorted.push_back(std::nullopt);
+      } else {
+        exec::VectorReader<Array<T>> reader(&decoded);
+        auto arrayView = reader[irow.first];
+        auto arrayEntry = arrayView[0];
+        if (!arrayEntry.has_value()) {
+          sorted.push_back({{{std::nullopt}}});
+        } else {
+          sorted.push_back({{{arrayEntry.value()}}});
+        }
+      }
+    }
+    return makeNullableArrayVector(sorted);
+  }
+
+  template <typename T>
+  void testCompareArrayFloats(TypePtr type, bool ascending, bool nullsFirst) {
+    // NOTE: set 'isJoinBuild' to true to enable nullable sort key in test.
+    auto rowContainer = makeRowContainer({type}, {type}, false);
+    auto lowest = std::numeric_limits<T>::lowest();
+    auto min = std::numeric_limits<T>::min();
+    auto max = std::numeric_limits<T>::max();
+    auto nan = std::numeric_limits<T>::quiet_NaN();
+    auto inf = std::numeric_limits<T>::infinity();
+
+    VectorPtr values = makeNullableArrayVector<T>(
+        {std::nullopt,
+         {{std::nullopt}},
+         {{max}},
+         {{inf}},
+         {{0.0}},
+         {{nan}},
+         {{lowest}},
+         {{min}}});
+    int numRows = values->size();
+
+    SelectivityVector allRows(numRows);
```

Comment:
The logic for storing all rows of a vector in a RowContainer repeats multiple times. Let's extract a helper method to reduce copy-paste.

```
  // A list of rows with unique zero-based indices.
  using IndexedRows = std::vector<std::pair<vector_size_t, char*>>;

  static std::vector<char*> store(
      DecodedVector& decoded,
      vector_size_t numRows,
      RowContainer& rowContainer) {
    std::vector<char*> rows(numRows);
    for (size_t i = 0; i < numRows; ++i) {
      rows[i] = rowContainer.newRow();
      rowContainer.store(decoded, i, rows[i], 0);
    }

    return rows;
  }

  static IndexedRows indexRows(const std::vector<char*>& rows) {
    IndexedRows indexedRows;
    indexedRows.reserve(rows.size());
    for (auto i = 0; i < rows.size(); ++i) {
      indexedRows.push_back({i, rows[i]});
    }
    return indexedRows;
  }


    int numRows = values->size();
    DecodedVector decoded(*values);
    auto rows = store(decoded, numRows, *rowContainer);
    auto indexedRows = indexRows(rows);
```